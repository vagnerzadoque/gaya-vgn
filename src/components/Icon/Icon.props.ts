import { IconName } from '@naturacosmeticos/natds-icons'
import { Color, Size } from '@naturacosmeticos/natds-themes'
import { BrandTypes } from '../../brandTypes/brandTypes'

export type IconSize = keyof Size
export type IconColor = keyof Color
export interface IconProps {
  /**
   * Removes the entire element from the accessibility tree
   * when the Icon is purely decorative
   * @default true
   */
  ariaHidden?: boolean

  brand?: BrandTypes

  /**
   * Use to define a string that labels the current element
   * if your icon is not decorative
   */
  ariaLabel?: string,

  /**
  * The Icon color
   * @default "primary"
   */
  color?: IconColor;

  /**
   * Override or extend the styles applied to the component
   */
  className?: string

  /**
   * Set the icon to be rendered.
   * Check all available names in [Icon Library](https://ds.natura.design/28db352be/p/94367e-icon-library/b/6154b9)
   * @default outlined-default-mockup
   */
  name: IconName;

  /**
   * @default 'img'
   */
  role?: 'img' | 'button',

  /**
  * The Icon size
   * @default "standard"
   */
  size?: IconSize;

  /**
   * Optional ID for testing
   */
  testID?: string,
}
