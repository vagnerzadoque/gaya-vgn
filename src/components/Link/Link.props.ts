export type LinkVariants = 'regular' | 'underline'
export type LinkPositions = 'left' | 'right'
type TextColors = 'default' | 'light'

export interface LinkProps {
  /**
   * Specify an optional className to be added to your Link
   */
  className?: string

  /**
   * Optional ID for testing
   */
  testID?: string;

  /**
   * The variant to use.
   */
  variant?: LinkVariants;

  /**
   * Specifies the URL of the page the link goes to.
   */
  href: React.AnchorHTMLAttributes<HTMLAnchorElement>['href'];

  /**
   * Specifies the language of the linked document.
   */
  hrefLang?: React.AnchorHTMLAttributes<HTMLAnchorElement>['hrefLang'];

  /**
   * The text to display inside the link.
   */
  text: string;

  /**
   * The icon to display together with the text.
   */
  textColor?: TextColors;

  /**
   * The target for the link.
   */
  target?: React.AnchorHTMLAttributes<HTMLAnchorElement>['target'];

  /**
   * The position of the icon.
   */
  iconPosition?: LinkPositions;

  /**
   * The icon to display together with the text.
   */
  IconComponent?: React.ReactElement;

}
