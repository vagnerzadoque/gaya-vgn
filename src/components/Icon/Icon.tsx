import React from 'react'
import { icons } from '@naturacosmeticos/natds-icons'
import { IconProps } from './Icon.props'
import styles from './Icon.styles'

export const checkIcon = (iconName: string): string => (iconName in icons ? iconName : 'outlined-default-mockup')

const Icon = React.forwardRef<HTMLElement, IconProps>(({
  ariaHidden = true,
  ariaLabel,
  className = '',
  color = 'highlight',
  brand,
  name,
  role = 'img',
  size = 'standard',
  testID = `icon-${name}`,
  ...props
}, ref) => {
  const { icon } = styles({ size, color, brand })
  const defaultIconName = `natds-icons-${checkIcon('outlined-default-mockup')}`
  const iconName = name ? `natds-icons-${checkIcon(name)}` : defaultIconName

  return (
    <i
      aria-hidden={ariaHidden}
      aria-label={ariaLabel}
      className={`${className} ${icon} natds-icons ${iconName} Icon-gaya`}
      data-testid={testID}
      ref={ref}
      role={role}
      {...props}
    />
  )
})

export default Icon
