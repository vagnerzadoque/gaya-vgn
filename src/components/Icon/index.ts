import Icon from './Icon'

export { Icon }
export { IconProps } from './Icon.props'
export default Icon
