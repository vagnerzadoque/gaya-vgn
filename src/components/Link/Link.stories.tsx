import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { Link, LinkProps } from '.'
import { Icon } from '../Icon'

const componentStatus = `
> The link component allows you to have a pattern for anchor text.

## Properties
| Property                  | Values                     |    Status           |
|---                        |                         ---|                  ---|
| **variant**               | regular, underline         | ✅ Available        |
| **href**                  | string href                | ✅ Available        |
| **hrefLang**              | string hrefLang            | ✅ Available        |
| **text**                  | string                     | ✅ Available        |
| **textColor**             | default, light             | ✅ Available        |
| **target**                | target ancor               | ✅ Available        |
| **iconPosition**          | left, right                | ✅ Available        |
| **IconComponent**         | component Icon             | ✅ Available        |


## Technical Usages Examples
`

const meta: Meta<LinkProps> = {
  title: 'Components/Link',
  component: Link,
  parameters: {
    componentSubtitle: '',
    docs: { description: { component: componentStatus } }
  }
};

export default meta;
type Story = StoryObj<LinkProps>;

export const Playground: Story = {
  args: {
    text: 'Link',
    href: 'https://'
  },
  render: (args) => <Link {...args} />
}

export const Variants: Story = {
  args: {
    href: 'https://'
  },
  render: (args) => (
    <>
      <Link {...args} variant="underline" text="underline variant example" />
      <Link {...args} variant="regular" text="regular variant example" />
    </>
  )
}

export const Colors: Story = {
  args: {
    href: 'https://'
  },
  render: (args) => (
    <>
      <Link {...args} text="default link color" />
      <Link {...args} textColor="light" text="light link color" />
    </>
  )
}

export const WithIcon: Story = {
  args: {
    href: 'https://'
  },
  render: (args) => (
    <>
      <Link
        {...args}
        text="right icon example"
        IconComponent={<Icon name="outlined-navigation-directionright-alt" color="link" />}
      />
      <Link
        {...args}
        text="left icon example"
        iconPosition="left"
        IconComponent={<Icon name="outlined-navigation-directionleft-alt" color="link" />}
      />
    </>
  )
}
