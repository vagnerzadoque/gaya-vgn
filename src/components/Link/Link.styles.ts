import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { LinkProps } from './Link.props'

type LinkStyleProps = Pick<LinkProps, 'iconPosition' | 'variant' | 'textColor' | 'IconComponent'>

const getIconPosition = () => ({ iconPosition }: LinkStyleProps) => (iconPosition === 'right' ? 'row' : 'row-reverse')
const getLabelMargin = (theme: Theme, side: string) => (
  { iconPosition }: LinkStyleProps
) => (iconPosition === side && theme.size.tiny)

const styles = createUseStyles((theme: Theme) => ({
  container: {
    alignItems: 'center',
    display: 'flex',
    flexDirection: getIconPosition(),
    justifyContent: 'center',
    width: 'fit-content',
    cursor: 'pointer'
  },
  link: {
    color: ({ textColor }: LinkStyleProps) => (textColor && textColor === 'default' ? theme.color.link : theme.color.lowEmphasis),
    fontFamily: [theme.link.label.primary.fontFamily, theme.link.label.fallback.fontFamily],
    fontSize: theme.link.label.fontSize,
    letterSpacing: theme.link.label.letterSpacing,
    lineHeight: theme.link.label.lineHeight,
    textDecoration: ({ variant }: LinkStyleProps) => (variant === 'underline' ? 'underline' : 'none'),
    marginLeft: getLabelMargin(theme, 'left'),
    marginRight: getLabelMargin(theme, 'right'),

    '&:hover': {
      textDecoration: ({ variant }: LinkStyleProps) => (variant === 'underline' ? 'none' : 'underline')
    }
  }
}))

export default styles
