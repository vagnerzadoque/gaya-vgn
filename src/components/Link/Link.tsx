import React from 'react'
import { LinkProps } from './Link.props'
import styles from './Link.styles'

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
  ({
    className = '',
    testID,
    variant = 'underline',
    href,
    hrefLang,
    text,
    target,
    iconPosition = 'right',
    IconComponent,
    textColor = 'default'
  }, ref) => {
    const { container, link } = styles({
      variant, iconPosition, textColor
    })

    return (
      <span className={container}>
        <a
          className={`${className} ${link} Link-gaya`}
          data-testid={testID}
          ref={ref}
          href={href}
          hrefLang={hrefLang}
          target={target}
        >
          {text}
        </a>
        {IconComponent && IconComponent}
      </span>
    )
  }
)

export default Link
