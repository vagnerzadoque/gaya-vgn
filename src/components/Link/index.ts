import Link from './Link'

export { Link }
export { LinkProps } from './Link.props'
export default Link
