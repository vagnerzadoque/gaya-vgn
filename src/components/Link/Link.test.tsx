import * as React from 'react'
import Link from '.'
import renderWithTheme from '../../helpers/renderWithTheme'
import { Icon } from '../Icon'

const defaultProps = {
  text: 'Link',
  href: 'https://'
}

describe('Link component', () => {
  it('should render correctly with default props', () => {
    const { styles, component } = renderWithTheme(<Link {...defaultProps} />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly when the variant is regular', () => {
    const { styles, component } = renderWithTheme(<Link {...defaultProps} variant="regular" />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly when the textColor is light', () => {
    const { styles, component } = renderWithTheme(<Link {...defaultProps} textColor="light" />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly with IconComponent rendered to the left', () => {
    const { styles, component } = renderWithTheme(
      <Link
        {...defaultProps}
        iconPosition="left"
        IconComponent={
          <Icon name="filled-default-mockup" />
        }
      />
    )

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly with IconComponent rendered to the right by default', () => {
    const { styles, component } = renderWithTheme(
      <Link
        {...defaultProps}
        IconComponent={
          <Icon name="filled-default-mockup" />
        }
      />
    )

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
})
