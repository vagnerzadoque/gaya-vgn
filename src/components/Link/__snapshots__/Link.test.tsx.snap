// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component should render correctly when the textColor is light 1`] = `
[
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #8c8c8c;
  text-decoration: underline;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4 Link-gaya"
        href="https://"
      >
        Link
      </a>
    </span>
  </div>,
]
`;

exports[`Link component should render correctly when the variant is regular 1`] = `
[
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #1267cf;
  text-decoration: none;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: underline;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4 Link-gaya"
        href="https://"
      >
        Link
      </a>
    </span>
  </div>,
]
`;

exports[`Link component should render correctly with IconComponent rendered to the left 1`] = `
[
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row-reverse;
}
.link-d1-0-2-4 {
  color: #1267cf;
  text-decoration: underline;
  margin-left: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}
.icon-0-2-5 {
  font-family: natds-icons;
  user-select: none;
  pointer-events: none;
}
.icon-d0-0-2-6 {
  color: #111111;
  font-size: 24px;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4 Link-gaya"
        href="https://"
      >
        Link
      </a>
      <i
        aria-hidden="true"
        class=" icon-0-2-5 icon-d0-0-2-6 natds-icons natds-icons-filled-default-mockup Icon-gaya"
        data-testid="icon-filled-default-mockup"
        role="img"
      />
    </span>
  </div>,
]
`;

exports[`Link component should render correctly with IconComponent rendered to the right by default 1`] = `
[
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #1267cf;
  text-decoration: underline;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}
.icon-0-2-5 {
  font-family: natds-icons;
  user-select: none;
  pointer-events: none;
}
.icon-d0-0-2-6 {
  color: #111111;
  font-size: 24px;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4 Link-gaya"
        href="https://"
      >
        Link
      </a>
      <i
        aria-hidden="true"
        class=" icon-0-2-5 icon-d0-0-2-6 natds-icons natds-icons-filled-default-mockup Icon-gaya"
        data-testid="icon-filled-default-mockup"
        role="img"
      />
    </span>
  </div>,
]
`;

exports[`Link component should render correctly with default props 1`] = `
[
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #1267cf;
  text-decoration: underline;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4 Link-gaya"
        href="https://"
      >
        Link
      </a>
    </span>
  </div>,
]
`;
